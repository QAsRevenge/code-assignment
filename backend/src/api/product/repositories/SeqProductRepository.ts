import {ProductType, Colour, Product} from '../../../_sequelize/models';
import {ProductDetailSchema} from '../schemas/responses/ProductDetailSchema';
import {ProductListSchemaArray} from '../schemas/responses/ProductListSchema';
import {ProductDetail} from '../types/responses/ProductDetail';
import {ProductList} from '../types/responses/ProductList';
import {Product as ProductInput} from '../types/validation/Product';
import {ProductRepository} from './ProductRepository';

export class SeqProductRepository implements ProductRepository {
  async createProduct(input: ProductInput): Promise<void> {
    const product: Product = await Product.create({
      name: input.name,
      productTypeId: input.productTypeId,
    });
    await product.$set('colours', input.colourIds);
  }
  async listProducts(productTypeId?: number): Promise<ProductList[]> {
    const where = productTypeId ? {productTypeId: productTypeId} : {};
    const products = await Product.findAll({
      where,
      attributes: ['id', 'name'],
      include: [
        {model: ProductType, attributes: ['productType']},
        {model: Colour, attributes: ['name'], through: {attributes: []}},
      ],
      order: [['createdAt', 'DESC']],
    });
    return ProductListSchemaArray.parse(products);
  }
  async getProductById(id: number): Promise<ProductDetail> {
    const product = await Product.findByPk(id, {
      include: [
        {
          model: ProductType,
          attributes: {
            exclude: ['createdAt', 'updatedAt'],
          },
        },
        {
          model: Colour,
          attributes: {exclude: ['createdAt', 'updatedAt']},
          through: {attributes: []},
        },
      ],
      attributes: {exclude: ['createdAt', 'updatedAt']},
    });
    return ProductDetailSchema.parse(product);
  }
  getProductTypes(): Promise<ProductType[]> {
    return ProductType.findAll({
      attributes: ['id', 'productType'],
    });
  }
  getProductColours(): Promise<Colour[]> {
    return Colour.findAll({attributes: ['id', 'name']});
  }
}
