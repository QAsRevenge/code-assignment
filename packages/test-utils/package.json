{
  "name": "test-utils",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "scripts": {
    "build": "pnpm exec tsc",
    "clean": "rm -rf dist",
    "lint": "eslint . --ext .ts",
    "test": "echo \"No tests specified\""
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "app-constants": "workspace:^",
    "sequelize": "^6.37.7",
    "sequelize-typescript": "^2.1.6",
    "app-db": "workspace:^"
  },
  "devDependencies": {
    "typescript": "^5.8.2"
  }
}
