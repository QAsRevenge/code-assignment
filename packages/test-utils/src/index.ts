export {SqlTestDb, SqlDbTestUtils} from './db/SqlDbTestUtils';
