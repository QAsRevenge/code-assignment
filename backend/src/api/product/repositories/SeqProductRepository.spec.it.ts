import {SqlDbTestUtils, SqlTestDb} from 'test-utils';
import {SeqProductRepository} from './SeqProductRepository';
import {
  describe,
  beforeAll,
  vi,
  it,
  expect,
  beforeEach,
  afterAll,
} from 'vitest';
import {
  Product,
  Colour,
  ProductType,
  ProductColour,
} from '../../../_sequelize/models';

describe('SeqProductRepository', () => {
  let productRepo: SeqProductRepository;
  let testDb: SqlTestDb;
  const timeout = 10 * 6000;

  beforeAll(async () => {
    vi.setSystemTime(new Date('2025-01-01'));
    testDb = await SqlDbTestUtils.createDb();
    testDb.sequelize.addModels([Product, Colour, ProductType, ProductColour]);
    productRepo = new SeqProductRepository();
  }, timeout);

  // This is very hacky and should be done in test-utils but for demonstration purposes..
  beforeEach(async () => {
    await testDb.sequelize.query(
      'TRUNCATE TABLE "data"."product_colours", "data"."products", "data"."colours", "data"."product_types" RESTART IDENTITY CASCADE;',
    );
  }, timeout);

  afterAll(async () => {
    await testDb.close();
  });

  it('should create a product', async () => {
    const colour1 = await Colour.create({name: 'Red', id: 1});
    const colour2 = await Colour.create({name: 'Blue', id: 2});
    const productType = await ProductType.create({productType: 'Shirt', id: 1});

    await productRepo.createProduct({
      name: 'Test Shirt',
      productTypeId: productType.id,
      colourIds: [colour1.id, colour2.id],
    });

    const products = await Product.findAll({
      include: [{model: Colour}, {model: ProductType}],
    });
    expect(products).toMatchSnapshot();
  });

  it('should list products with no productTypeId specified', async () => {
    const colour1 = await Colour.create({name: 'Red'});
    const colour2 = await Colour.create({name: 'Blue'});
    const productType1 = await ProductType.create({productType: 'Shirt'});
    const productType2 = await ProductType.create({productType: 'Pants'});

    const product1 = await Product.create({
      name: 'Test Shirt 1',
      productTypeId: productType1.id,
      createdAt: new Date('2024-01-01'),
    });
    await product1.$set('colours', [colour1.id]);

    const product2 = await Product.create({
      name: 'Test Pants 1',
      productTypeId: productType2.id,
      createdAt: new Date('2024-06-01'),
    });
    await product2.$set('colours', [colour2.id]);

    const products = await productRepo.listProducts();
    expect(products).toMatchSnapshot();
  });

  it('should list products with productTypeId specified', async () => {
    const colour1 = await Colour.create({name: 'Red'});
    const colour2 = await Colour.create({name: 'Blue'});
    const productType1 = await ProductType.create({productType: 'Shirt'});
    const productType2 = await ProductType.create({productType: 'Pants'});

    const product1 = await Product.create({
      name: 'Test Shirt 1',
      productTypeId: productType1.id,
      createdAt: new Date('2024-01-01'),
    });
    await product1.$set('colours', [colour1.id]);

    const product2 = await Product.create({
      name: 'Test Pants 1',
      productTypeId: productType2.id,
      createdAt: new Date('2024-06-01'),
    });
    await product2.$set('colours', [colour2.id]);

    const products = await productRepo.listProducts(productType1.id);
    expect(products).toMatchSnapshot();
  });

  it('should get product by id', async () => {
    const colour1 = await Colour.create({name: 'Red'});
    const colour2 = await Colour.create({name: 'Blue'});
    const productType1 = await ProductType.create({productType: 'Shirt'});

    const product1 = await Product.create({
      name: 'Test Shirt 1',
      productTypeId: productType1.id,
      createdAt: new Date('2024-01-01'),
    });
    await product1.$set('colours', [colour1.id, colour2.id]);

    const product = await productRepo.getProductById(product1.id);
    expect(product).toMatchSnapshot();
  });

  it('should get all product types', async () => {
    await ProductType.create({productType: 'Shirt'});
    await ProductType.create({productType: 'Pants'});

    const productTypes = await productRepo.getProductTypes();
    expect(productTypes).toMatchSnapshot();
  });

  it('should get all product colours', async () => {
    await Colour.create({name: 'Red'});
    await Colour.create({name: 'Blue'});

    const colours = await productRepo.getProductColours();
    expect(colours).toMatchSnapshot();
  });
});
