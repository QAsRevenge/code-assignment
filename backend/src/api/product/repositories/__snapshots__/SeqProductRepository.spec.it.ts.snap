// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeqProductRepository > should create a product 1`] = `
[
  {
    "colours": [
      {
        "ProductColour": {
          "colourId": 1,
          "createdAt": 2025-01-01T00:00:00.000Z,
          "id": 1,
          "productId": 1,
          "updatedAt": 2025-01-01T00:00:00.000Z,
        },
        "createdAt": 2025-01-01T00:00:00.000Z,
        "id": 1,
        "name": "Red",
        "updatedAt": 2025-01-01T00:00:00.000Z,
      },
      {
        "ProductColour": {
          "colourId": 2,
          "createdAt": 2025-01-01T00:00:00.000Z,
          "id": 2,
          "productId": 1,
          "updatedAt": 2025-01-01T00:00:00.000Z,
        },
        "createdAt": 2025-01-01T00:00:00.000Z,
        "id": 2,
        "name": "Blue",
        "updatedAt": 2025-01-01T00:00:00.000Z,
      },
    ],
    "createdAt": 2025-01-01T00:00:00.000Z,
    "id": 1,
    "name": "Test Shirt",
    "productType": {
      "createdAt": 2025-01-01T00:00:00.000Z,
      "id": 1,
      "productType": "Shirt",
      "updatedAt": 2025-01-01T00:00:00.000Z,
    },
    "productTypeId": 1,
    "updatedAt": 2025-01-01T00:00:00.000Z,
  },
]
`;

exports[`SeqProductRepository > should get all product colours 1`] = `
[
  {
    "id": 1,
    "name": "Red",
  },
  {
    "id": 2,
    "name": "Blue",
  },
]
`;

exports[`SeqProductRepository > should get all product types 1`] = `
[
  {
    "id": 1,
    "productType": "Shirt",
  },
  {
    "id": 2,
    "productType": "Pants",
  },
]
`;

exports[`SeqProductRepository > should get product by id 1`] = `
{
  "colours": [
    "Red",
    "Blue",
  ],
  "id": 1,
  "name": "Test Shirt 1",
  "productType": "Shirt",
}
`;

exports[`SeqProductRepository > should list products with no productTypeId specified 1`] = `
[
  {
    "colours": "Blue",
    "id": 2,
    "name": "Test Pants 1",
    "productType": "Pants",
  },
  {
    "colours": "Red",
    "id": 1,
    "name": "Test Shirt 1",
    "productType": "Shirt",
  },
]
`;

exports[`SeqProductRepository > should list products with productTypeId specified 1`] = `
[
  {
    "colours": "Red",
    "id": 1,
    "name": "Test Shirt 1",
    "productType": "Shirt",
  },
]
`;
